### PR TITLE
 feat: Add autofocus for `MenuItemButton`

### LINKED ISSUE #139395: Improve and fix MenuAnchor/MenuItemButton UX for desktop
MenuAnchor for Material 3 is great for contextual menus but there are some minor issues related to accessibility. Specifically, keyboard navigation does not auto-select an element when opening the menu. This causes some UX confusion where the user has to navigate further to access the menu he or she just opened, which in my case is not user friendly.

Setup:
- Material 3
- Windows 10 22H2
- Flutter 3.16.2

Relevant PR:
- [ ] https://github.com/flutter/flutter/pull/139396 MenuItemButton doesn't have an `autofocus` attribute, thereby making it harder to navigate with the keyboard. 

Thanks Flutter Team!

### PR DESCRIPTION
**Pull request by Fernthedev**
MenuAnchor for Material 3 is great for contextual menus but there are some minor issues related to accessibility.
This PR aims to close the gap by adding `autofocus` to the menu item button so keyboard navigation is more intuitive. Otherwise, it becomes a mess to navigate through just keyboards.

Partially resolves #139395

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

## Additional Notes

I should mention, I have not written tests for this due to it's trivial nature. I also lack the experience of writing Flutter tests in general, so if someone feels inclined to take over this PR and add it they're welcome to.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

**Comment by gspencergoog** [comment]
Thanks for the PR!

In order to review of this PR, however, we'll need you to add some tests, as the bot message describes above.

**Comment by Fernthedev** [comment]
> Hi @Fernthedev would you like to continue working on this PR and add some tests?

Hi, yes! I apologize for my absence since I've been very busy and forgot about this.

I don't have any experience with test development in Flutter so I'm not sure how to do so. If you're offering to add them yourself or take over the PR, I endorse it. If not I'll try to do it this Friday as I think I'll have time by then.

Thanks for your interest!

**Comment by Fernthedev** [comment]
Hi @Piinks (hope you don't mind the ping) I have added the tests as requested. 

However, I'd like to note that during testing, I noticed that `MenuBar` does not seem to respond to `autofocus` at all. Or any sort of  `Focus` in general from what I could tell. The `primaryFocus` would always be the modal route.

I hope the test adequately conforms to the Flutter repo's requirements but in the case it does not feel free to ping me.

**Review by Piinks — COMMENTED** [review]
Thanks for adding tests! I'll defer to @gspencergoog for this, I don't know if the focus issue you mentioned in the menu is intentional or a separate issue.

File: packages/flutter/test/material/menu_anchor_test.dart [hunk 1]
```diff
@@ -2227,6 +2227,48 @@ void main() {
 
       expect(find.text('leadingIcon'), findsOneWidget);
     });
+    testWidgets('autofocus is used when set and widget is enabled',
+        (WidgetTester tester) async {
+
+      listenForFocusChanges();
+
+      // Note: This DOES NOT WORK WITH MenuBar!
+      // Focus does not work under any circumstance with MenuBar!
```
Comment (Piinks) [inline packages/flutter/test/material/menu_anchor_test.dart:2236]: Can you share some context on this comment? Is this meant to be this way?

**Review by Piinks — COMMENTED** [review]
Hi @Fernthedev, this came PR came up in PR triage this week. Can you let us know that status of this change? It looks like you mentioned this is a partial fix, is there something missing here? Thanks!

**Comment by Fernthedev** [comment]
> Hi @Fernthedev, this came PR came up in PR triage this week. Can you let us know that status of this change? It looks like you mentioned this is a partial fix, is there something missing here? Thanks!

Hey thanks! So basically when I originally wrote this I was listing 3 specific issues in the original issue. That's no longer the case, this PR addresses the issue specifically mentioned in the original GitHub post.

Here's a video referencing the problem I tried to resolve with this. https://github.com/flutter/flutter/issues/139395#issuecomment-1847830778

I apologize for the confusion, I wasn't thinking well when I wrote these posts. 😅

tl;dr this should resolve the problem where opening the menu would no select an item and therefore require extra navigation steps. Example being opening the menu with keyboard and attempt an action.

File: packages/flutter/test/material/menu_anchor_test.dart [hunk 2]
```diff
@@ -2227,6 +2227,48 @@ void main() {
 
       expect(find.text('leadingIcon'), findsOneWidget);
     });
+    testWidgets('autofocus is used when set and widget is enabled',
```
Comment (QuncCccccc) [inline packages/flutter/test/material/menu_anchor_test.dart:2230]:
```suggestion

    testWidgets('autofocus is used when set and widget is enabled',
```

File: packages/flutter/test/material/menu_anchor_test.dart (hunk 1 above)
Comment (QuncCccccc) [reply packages/flutter/test/material/menu_anchor_test.dart:2236]: +1

**Review by QuncCccccc — COMMENTED** [review]
Thanks for the contribution! Could you provide some more details about the question that @Piinks asked?

**Comment by Piinks** [comment]
Hi @Fernthedev would you like to continue working on this PR and address the above feedback?

**Comment by Fernthedev** [comment]
Hi, I left a response to a review left by @QuncCccccc. Did I miss something else you needed clarified?

File: packages/flutter/test/material/menu_anchor_test.dart (hunk 1 above)
Comment (QuncCccccc) [reply packages/flutter/test/material/menu_anchor_test.dart:2236]: Got it, thanks for the explanation! Can we remove this comment if it's not related to this PR? If this is a bug on MenuBar, we probably should create an issue ticket for it:)

Comment (Fernthedev) [reply packages/flutter/test/material/menu_anchor_test.dart:2236]: Absolutely! I'll look into making an issue/ticket for clarification whether this is intended or not. 

Thanks for your patience 😅

Comment (QuncCccccc) [reply packages/flutter/test/material/menu_anchor_test.dart:2236]: No problem! Once we remove the comments here and add one line between unit tests([here](https://github.com/flutter/flutter/pull/139396#discussion_r1635341777)), I'll ask for a second review, then this PR is good to go! Thanks again for the contribution:)

File: packages/flutter/test/material/menu_anchor_test.dart [hunk 3]
```diff
@@ -2228,6 +2228,47 @@ void main() {
       expect(find.text('leadingIcon'), findsOneWidget);
     });
 
+    testWidgets('autofocus is used when set and widget is enabled',
+        (WidgetTester tester) async {
+
+      listenForFocusChanges();
+
+
```
Comment (QuncCccccc) [inline packages/flutter/test/material/menu_anchor_test.dart:2236]: nit: remove the extra line
```suggestion
```

**Review by QuncCccccc — APPROVED** [review]
LGTM! Thanks for the contribution:)

**Review by MitchellGoodwin — APPROVED** [review]
LGTM!

File: packages/flutter/test/material/menu_anchor_test.dart [hunk 4]
```diff
@@ -2229,6 +2229,46 @@ void main() {
...
+        MaterialApp(
+          home: Material(
+            child: Column(
+              children: <Widget>[
+                MenuAnchor(
+                  controller: controller,
+                  menuChildren: <Widget>[
+                    MenuItemButton(
+                      autofocus: true,
+                      // required for clickability
```
Comment (justinmc) [inline packages/flutter/test/material/menu_anchor_test.dart:2247]: Capital and period: "Required for clickability."

File: packages/flutter/test/material/menu_anchor_test.dart [hunk 5]
```diff
@@ -2229,6 +2229,46 @@ void main() {
...
+          ),
+        ),
+      );
+
+      controller.open();
+      await tester.pump();
+
+      expect(controller.isOpen, equals(true));
+
+      expect(focusedMenu, equals('MenuItemButton(Text("Menu 0"))'));
```
Comment (justinmc) [inline packages/flutter/test/material/menu_anchor_test.dart:2269]: Nit: Maybe use `TestMenu.mainMenu0.label` instead of the string here.

**Review by justinmc — APPROVED** [review]
LGTM with nits :+1: Thanks for adding this!

**Comment by justinmc** [comment]
Thanks for the quick fixes.
